### ISSUE TITLE
Milestones, quests, queststeps, and objectives

### ISSUE THREAD

**Issue by k20stitch** [Bungie-net/api#233]
One would assume that if a milestone is of type... 3 for example, all type 3 milestones would be standardized. but some of display data and some don't, Some milestone names are different than their quest names (Hotspot vs flashpoint on ::insert planet here::). Then you have the ikora missions which have a milestone name, but the individual quests are not named.

If someone is trying to display the weeks milestone data, like you guys do in the app, what should be we using? The most reliable i've found is the questItem definition (the only one that doesn't work is the Ikora meditations).

**Comment by floatingatoll** [comment]
Can you be more specific about precisely which milestones and quests are
giving you this trouble? They may require individual attention due to
classification and/or bugs and/or etc and it would be invaluable to know
which ones you've found work, which don't, etc.

**Comment by k20stitch** [comment]
I'm trying to mimic what you see in the director or milestones in the app.

Take the type 3 (weekly) milestones for example.  I query what milestones (character progression endpoint) are available for a given character and I'm given a pretty standard response... the milestone and hash along with some available quests and the start/end times.

Next i get the milestone definition and this is where it gets ugly. For the nightfall, call to arms, raid, trials of the nine, i get display properties which includes the name description and icon. But not for the "hotspot"/"flash points". The clan XP milestone is also set to false for display in explorer but it is clearly shown in the director (unlike the meditations and clan progression milestones). The other interesting thing here is that the names and descriptions for the milestone display properties do not match what's shown in the directory or app. Because of this, i decided to go down the milestone chain and get definitions for the next best things, the "quests" within the milestone.

Most quests had the display properties I'm looking for and the descriptions matched the information shown in the director. (compare raid milestone definition to raid quest definition). The hotspot, is also now properly named to "flash point on ::insert planet here::".. so great, i'll use the quest definitions instead... NOPE.

Now when you look at the Ikora mediation milestone that did have a milestone display properties,  there are no display properties in the quest definition.

I've even tried using the activity definition, objective definitions, and even the conceptual activity definition. but none of them are ever standard across all of the weekly type 3 milestones.  I'd like to develop my app so that when new milestones are available, my app just programmatically displays them instead of all this ifthenelse BS where I'll need to know before hand which milestone information i need to grab for each milestone.

Or perhaps i'm just going about this whole thing wrong.

**Comment by hskrasek** [comment]
@k20stitch Might not help much, but I threw together a Slackbot for my group that posts the weekly milestones every tuesday and kinda had the same issues you were having, but to get output that looks like this:

![screenshot 2017-10-31 16 17 09](https://user-images.githubusercontent.com/787487/32249417-f59526f0-be56-11e7-8d73-e3a9d79dc7c5.png)

I had to make a system that does some custom parsing for each milestone/milestone type. Not sure if you are familiar with PHP, but you can see the differences in each class here: https://github.com/hskrasek/reinbot/tree/master/app/Services/Destiny/Transformers

But yea everything is a bit scattered just to piece all the information together. My approach is based off of milestone hashes, and I'm sure has like no fault tolerance. But it'd be a good place to start

**Comment by k20stitch** [comment]
Yea this is what I’m trying to avoid. It requires that you already know what the milestones are going to be if you need custom parsing. 

Sent via smoke signals

**Comment by vthornheart-bng** [comment]
Yes, unfortunately milestones currently have very inconsistent display information whose quality really isn't that great.  The rule of thumb that we apply on our site and in our apps is to use the most specific description we can find, if it is populated.  So we first try to show the Quest's information as the Milestone.  If it doesn't have display data populated, we'll fall back to showing the milestone-level display properties.  We use the displayProperties under the "quests" property in the Milestone definition, rather than the quest item's displayProperties.  Unfortunately, something that both the game and our data (as a result) is missing is more useful information about *what* the Milestone is.

This is an experience that I'm hoping to improve over time, but for now both you and we have to deal with this inconsistent data source unfortunately.  But if you use that algorithm of using the Milestone quests' displayProperties and falling back on the Milestone display properties when the quest has no display properties, that will at least get you to show something that should have a baseline level of usefulness.  Eventually I'll be attempting to get better text into those fields, even if we have to rig it up manually: but it'll always end up being a pattern of checking the quest's display properties first and falling back on the Milestone, because some Milestones will inherently have very different descriptions based on the active quest.

When it comes to the "display in Explorer" property, that's what we use to determine whether or not we'll display that milestone in the "Live Events" section of the frontpage of the website: it's got no bearing on whether or not the milestone gets shown in the director.

**Comment by floatingatoll** [comment]
I nominate this comment to be included in the docs somehow, because it's
critically valuable for anyone parsing milestones.

**Comment by vthornheart-bng** [comment]
100% agreed, very good call.  I'll put the gist of this into the docs while we're talking about it for the next time I update the docs.

**Comment by vthornheart-bng** [comment]
Here we go, for next time we update docs:

/// There are some important caveats to understand about how we currently render Milestones and their deficiencies.  The game
	/// currently doesn't have any content that actually tells you oughtright *what* the Milestone is: that is to say, what you'll
	/// be doing.  The best we get is either a description of the overall Milestone, or of the Quest that the Milestone is having
	/// you partake in: which is usually something that assumes you already know what it's talking about, like "Complete 5 Challenges".
	/// 5 Challenges for what?  What's a challenge?  These are not questions that the Milestone data will answer for you unfortunately.
	/// 
	/// This isn't great, and in the future I'd like to add some custom text to give you more contextual information to pass on to
	/// your users.  But for now, you can do what we do to render what little display info we do have:
	/// 
	/// Start by looking at the currently active quest (ideally, you've fetched DestinyMilestone or DestinyPublicMilestone data
	/// from the API, so you know the currently active quest for the Milestone in question).  Look up the Quests property in the
	/// Milestone Definition, and check if it has display properties.  If it does, show that as the description of the Milestone.
	/// If it doesn't, fall back on the Milestone's description.
	/// 
	/// This approach will let you avoid, whenever possible, the even less useful (and sometimes nonexistant) milestone-level names
	/// and descriptions.

**Comment by k20stitch** [comment]
Thank you for the super informative post! Glad this got the attention it deserved.
